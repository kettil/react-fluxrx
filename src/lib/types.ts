import { ComponentType } from 'react';
import { Observable, Subscription } from 'rxjs';

//
// Action
//

export type actionType<PAYLOAD = any> = {
  type: string | symbol;
  payload: PAYLOAD;
};

export type actionSubjectType<PAYLOAD = any> = actionType<PAYLOAD> | Observable<actionType<PAYLOAD>>;

export type withChildrenType<Props extends {}> = Props & { readonly children: React.ReactNode };

//
// Connect
//

export type connectedType<Props, MapState, MapDispatch> = (
  Element: ComponentType<mergedObjects<Props, MapState, MapDispatch>>,
) => ComponentType<Props>;

export type mapStateToPropsType<State, Props, MapState> = (state: State, ownProps: Props) => MapState;

export type mapDispatchToPropsType<Props, MapDispatch> = (dispatch: storeDispatchType, ownProps: Props) => MapDispatch;

export type mergePropsType<MapState, MapDispatch, Props> = (
  stateProps: MapState,
  dispatchProps: MapDispatch,
  ownProps: Props,
) => mergedObjects<Props, MapState, MapDispatch>;

export type wrappedComponentType<Props, MapState, MapDispatch> =
  | React.ComponentType<mergedObjects<Props, MapState, MapDispatch>>
  | React.ReactElement<mergedObjects<Props, MapState, MapDispatch>>;

//
// Store
//

export type storeDispatchType<PAYLOAD = any> = (action: actionSubjectType<PAYLOAD>) => void;

export type storeSubscribeType<State> = (state: State) => void;

export type storeType<State, PAYLOAD = any> = {
  subscribe: (next: storeSubscribeType<State>) => Subscription;
  dispatch: storeDispatchType<PAYLOAD>;
  getState: () => State;
};

export type storeSetStateType<State> = (state: State) => void;

export type storeErrorHandlerType<State> = (err: any, dispatch: storeDispatchType, state: State) => void;

//
// Middleware
//

export type middlewareType<State> = {
  init?: middlewareInitType<State>;
  action?: middlewareActionType<State>;
  error?: middlewareErrorType<State>;
};

export type middlewareInitType<State, PAYLOAD = any> = (
  state: State,
  dispatch: storeDispatchType<PAYLOAD>,
  setState: storeSubscribeType<State>,
  updateDirectly: storeSubscribeType<State>,
) => void;

export type middlewareActionType<State, PAYLOAD = any> = (
  action: actionType,
  state: State,
  dispatch: storeDispatchType<PAYLOAD>,
  reducer: reducerType<State>,
) => actionSubjectType;

export type middlewareErrorType<State> = storeErrorHandlerType<State>;

//
// Reducer
//

export type reducersType<State, ACTION extends actionType = any> = {
  [K in keyof State]: reducerType<State[K], ACTION>;
};

export type reducerType<State, ACTION extends actionType = any> = (state: State | undefined, action: ACTION) => State;

//
// Helper
//

export type mergedObjects<T1, T2, T3> = T3 & objectExcludeKeysType<T2, T3> & objectExcludeKeysType<T1, T2 & T3>;

export type objectExcludeKeysType<T1, T2> = { [K in Exclude<keyof T1, keyof T2>]: T1[K] };
