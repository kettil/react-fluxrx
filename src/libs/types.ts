import { Subscription as SubscriptionReact, SubscriptionConfig } from 'create-subscription';
import { Observable, ObservableInput, Subscription } from 'rxjs';

//////////////////////
//
// INDEX
//

/**
 *
 */
export type optionsType<S> = {
  createContext?: createContextType<storeType<S>>;
  createStore?: createStoreType<S>;
  createConnect?: createConnectType<S>;

  mapStateToProps?: mapStateToPropsType<S, any, any>;
  mapDispatchToProps?: mapDispatchToPropsType<any, any>;
  mergeProps?: mergePropsType<any, any, any>;

  mapStateToPropsWithCacheFactory?: mapStateToPropsWithCacheFactoryType<S, any, any>;
  mapDispatchToPropsWithCacheFactory?: mapDispatchToPropsWithCacheFactoryType<any, any>;
  mergePropsWithCacheFactory?: mergePropsWithCacheFactoryType<any, any, any>;
  propsFactory?: propsFactoryType<S, any, any, any>;

  createElementWithoutSubscription?: createElementWithoutSubscriptionType<S, any, any, any>;
  createElementWithSubscription?: createElementWithSubscriptionType<S, any, any, any>;
  createSubscriptionWrapper?: createSubscriptionWrapperType<S, any, any, any>;
  createSubscription?: createSubscriptionType<S, any, any, any>;

  areStatesEqual?: equalType<S, S>;
  arePropsEqual?: equalType<any, any>;
  areMappedEqual?: equalType<any, any>;
  areDispatchedEqual?: equalType<any, any>;

  middlewareManager?: middlewareManagerType;
  middlewareHandler?: middlewareHandlerType;

  reducerHandler?: reducerHandlerType<S>;
  errorStoreHandler?: errorHandlerType;

  actionFilter?: actionValidatorType;
  actionFlat?: actionFlatType;
  actionError?: actionErrorType;
};

/**
 *
 */
export type optionsConnectType<S, P, MS, MD> = {
  mapStateToProps: mapStateToPropsType<S, P, MS>;
  mapDispatchToProps: mapDispatchToPropsType<P, MD>;
  mergeProps: mergePropsType<MS, MD, P>;

  mapStateToPropsWithCacheFactory: mapStateToPropsWithCacheFactoryType<S, P, MS>;
  mapDispatchToPropsWithCacheFactory: mapDispatchToPropsWithCacheFactoryType<P, MD>;
  mergePropsWithCacheFactory: mergePropsWithCacheFactoryType<P, MS, MD>;
  propsFactory: propsFactoryType<S, P, MS, MD>;

  createElementWithoutSubscription: createElementWithoutSubscriptionType<S, P, MS, MD>;
  createElementWithSubscription: createElementWithSubscriptionType<S, P, MS, MD>;
  createSubscriptionWrapper: createSubscriptionWrapperType<S, P, MS, MD>;
  createSubscription: createSubscriptionType<S, P, MS, MD>;

  areStatesEqual: equalType<S, S>;
  arePropsEqual: equalType<P, P>;
  areMappedEqual: equalType<MS, MS>;
  areDispatchedEqual: equalType<MD, MD>;
};

/**
 *
 */
export type optionsStoreType<S> = {
  middlewareManager: middlewareManagerType;
  middlewareHandler: middlewareHandlerType;

  reducerHandler: reducerHandlerType<S>;
  errorStoreHandler: errorHandlerType;

  actionFilter: actionValidatorType;
  actionFlat: actionFlatType;
  actionError: actionErrorType;
};

/**
 *
 */
export type createContextType<T> = (
  defaultValue: T,
  calculateChangedBits?: ((prev: T, next: T) => number) | undefined,
) => React.Context<T>;

/**
 *
 */
export type equalType<A = any, B = any> = (a: A, b: B) => boolean;

/**
 *
 */
export type logType = (a: any) => void;

//////////////////////
//
// CONNECT
//

export type createConnectType<S> = (
  Consumer: React.ComponentType<React.ConsumerProps<storeType<S>>>,
  options: optionsConnectType<S, any, any, any>,
) => <P, MS, MD>(
  mapStateToProps?: mapStateToPropsType<S, P, MS> | null,
  mapDispatchToProps?: mapDispatchToPropsType<P, MD>,
  mergeProps?: mergePropsType<MS, MD, P>,
) => connectedType<P, MS, MD>;

/**
 *
 */
export type createElementWithoutSubscriptionType<S, P, MS, MD> = (
  Consumer: React.ComponentType<React.ConsumerProps<storeType<S>>>,
  options: optionsConnectType<S, P, MS, MD>,
) => (Element: React.ComponentType<propsMergeReturnType<P, MS, MD>>) => React.ComponentType<P>;

/**
 *
 */
export type createElementWithSubscriptionType<S, P, MS, MD> = (
  Consumer: React.ComponentType<React.ConsumerProps<storeType<S>>>,
  options: optionsConnectType<S, P, MS, MD>,
) => (Element: React.ComponentType<propsMergeReturnType<P, MS, MD>>) => React.ComponentType<P>;

/**
 *
 */
export type subscriptionType<S, P> = {
  store: storeType<S>;
  ownProps: P;
};

/**
 *
 */
export type createSubscriptionType<S, P, MS, MD> = (
  config: SubscriptionConfig<subscriptionType<S, P>, propsMergeReturnType<P, MS, MD>>,
) => SubscriptionReact<subscriptionType<S, P>, propsMergeReturnType<P, MS, MD>>;

/**
 *
 */
export type createSubscriptionWrapperType<S, P, MS, MD> = (
  createSubscription: createSubscriptionType<S, P, MS, MD>,
  merge: propsMergeType<S, P, MS, MD>,
) => SubscriptionReact<subscriptionType<S, P>, propsMergeReturnType<P, MS, MD>>;

/**
 *
 */
export type connectedType<P, MS, MD> = (
  Element: React.ComponentType<propsMergeReturnType<P, MS, MD>>,
) => React.ComponentType<P>;

/**
 *
 */
export type mapStateToPropsType<S, P, MS> = (state: S, ownProps: P) => MS;

/**
 *
 */
export type mapDispatchToPropsType<P, MD> = (dispatch: dispatchType, ownProps: P) => MD;

/**
 *
 */
export type mergePropsType<MS, MD, P> = (
  stateProps: MS,
  dispatchProps: MD,
  ownProps: P,
) => propsMergeReturnType<P, MS, MD>;

/**
 * @todo change type name
 */
export type propsMergeType<S, P, MS, MD> = (nextState: S, nextProps: P) => propsMergeReturnType<P, MS, MD>;

/**
 *
 */
export type propsMergeReturnType<T1, T2, T3> = T3 & objectExcludeKEysType<T2, T3> & objectExcludeKEysType<T1, T2 & T3>;

/**
 *
 */
export type objectExcludeKEysType<T1, T2> = { [K in Exclude<keyof T1, keyof T2>]: T1[K] };

//////////////////////
//
// FACTORY
//

/**
 *
 */
export type mapStateToPropsWithCacheFactoryType<S, P, MS> = (
  mapStateToProps: mapStateToPropsType<S, P, MS>,
  areStatesEqual: equalType<S, S>,
  arePropsEqual: equalType<P, P>,
) => mapStateToPropsWithCacheType<S, P, MS>;

/**
 *
 */
export type mapStateToPropsWithCacheType<S, P, MS> = (nextState: S, nextProps: P, hasDependsOnProps: boolean) => MS;

/**
 *
 */
export type mapDispatchToPropsWithCacheFactoryType<P, MD> = (
  mapDispatchToProps: mapDispatchToPropsType<P, MD>,
  arePropsEqual: equalType<P, P>,
) => mapDispatchToPropsWithCacheType<P, MD>;

/**
 *
 */
export type mapDispatchToPropsWithCacheType<P, MD> = (
  dispatch: dispatchType,
  nextProps: P,
  hasDependsOnProps: boolean,
) => MD;

/**
 *
 */
export type mergePropsWithCacheFactoryType<P, MS, MD> = (
  mergeProps: mergePropsType<MS, MD, P>,
  arePropsEqual: equalType<P, P>,
  areMappedEqual: equalType<MS, MS>,
  areDispatchedEqual: equalType<MD, MD>,
) => mergePropsWithCacheType<MS, MD, P>;

/**
 *
 */
export type mergePropsWithCacheType<MS, MD, P> = (
  stateMapped: MS,
  dispatchMapped: MD,
  props: P,
) => propsMergeReturnType<P, MS, MD>;

/**
 *
 */
export type propsFactoryType<S, P, MS, MD> = (
  mapStateToPropsWithCache: mapStateToPropsWithCacheType<S, P, MS>,
  mapDispatchToPropsWithCache: mapDispatchToPropsWithCacheType<P, MD>,
  mergePropsWithCache: mergePropsWithCacheType<MS, MD, P>,
  mapStateToProps: mapStateToPropsType<S, P, MS>,
  mapDispatchToProps: mapDispatchToPropsType<P, MD>,
) => propsFactoryDispatchType<S, P, MS, MD>;

/**
 *
 */
export type propsFactoryDispatchType<S, P, MS, MD> = (dispatch: dispatchType) => propsFactoryUpdateType<S, P, MS, MD>;

/**
 *
 */
export type propsFactoryUpdateType<S, P, MS, MD> = (nextState: S, nextProps: P) => propsMergeReturnType<P, MS, MD>;

//////////////////////
//
// MIDDLEWARE
//

/**
 *
 */
export type middlewareActionType = (action: actionType, dispatch: dispatchType) => actionSubjectType;

/**
 *
 */
export type middlewareManagerType = (
  dispatch: dispatchType,
  middlewares: middlewareActionType[],
  actionValidator: actionValidatorType,
  middlewareHandler: middlewareHandlerType,
) => (source$: Observable<actionType>) => Observable<actionType>;

/**
 *
 */
export type middlewareHandlerType = (
  dispatch: dispatchType,
  middlewares: middlewareActionType,
  actionValidator: actionValidatorType,
  source$: Observable<actionType>,
) => Observable<actionType>;

//////////////////////
//
// REDUCERS
//

/**
 *
 */
export type reducerHandlerType<S> = (reducer: reducerType<S>) => reducerType<S>;

/**
 *
 */
export type reducersType<S> = {
  [key: string]: reducerType<S>;
};

/**
 *
 */
export type reducerType<S> = (state: S | undefined, action: actionType) => S;

//////////////////////
//
// STORE
//

/**
 *
 */
export type createStoreType<S> = (
  reducer: reducerType<S>,
  init: S,
  middleware: middlewareActionType[],
  options: optionsStoreType<S>,
) => storeType<S>;

/**
 *
 */
export type actionType = {
  type: string;
  payload: any;
};

/**
 *
 */
export type actionSubjectType = actionType | Observable<actionType>;

/**
 *
 */
export type dispatchType = (action: actionSubjectType) => void;

/**
 *
 */
export type subscribeType<S> = (state: S) => void;

/**
 *
 */
export type storeType<S> = {
  subscribe: (next: subscribeType<S>) => Subscription;
  dispatch: dispatchType;
  getState: () => S;
};

/**
 *
 */
export type actionValidatorType = (action: any) => action is actionType;

/**
 *
 */
export type actionFlatType = (action: actionSubjectType) => Observable<actionType>;

/**
 *
 */
export type actionErrorType = (
  errorHandler: errorHandlerType,
  dispatch: dispatchType,
) => (error: any, action$: Observable<actionType>) => ObservableInput<actionType>;

/**
 *
 */
export type errorHandlerType = (error: any, dispatch: dispatchType) => void;
