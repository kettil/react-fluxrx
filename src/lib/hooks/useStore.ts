import { Context, useContext } from 'react';
import { StoreType } from '../types';

export const createStoreHook = <State>(context: Context<StoreType<State, any>>) => {
  const useStore = () => useContext(context);

  return useStore;
};
