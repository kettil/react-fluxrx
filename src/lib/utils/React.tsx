import { FunctionComponent, memo } from 'react';

export const hocOptimize = <C, W>(Component: FunctionComponent<C>, Wrapper: FunctionComponent<W>) => {
  const displayName = Component.displayName || Component.name;

  // istanbul ignore else
  if (displayName) {
    Wrapper.displayName = displayName;
  }

  return memo(Wrapper);
};
