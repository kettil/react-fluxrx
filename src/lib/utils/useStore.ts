import { useContext, useEffect, useMemo, useState } from 'react';
import selector from '../selector';
import { mapDispatchToPropsType, mapStateToPropsType, storeType } from '../types';
import { isStrictEqual } from './connect';

type Store<State, InnerProps, OuterProps> = {
  merge: InnerProps;
  state: State;
  inner: InnerProps;
  outer: OuterProps;
};

export const useStore = <State, InnerProps, OuterProps, MapState, MapDispatch>(
  context: React.Context<storeType<State>>,
  mapStateToProps: mapStateToPropsType<State, OuterProps, MapState>,
  mapDispatchToProps: mapDispatchToPropsType<OuterProps, MapDispatch>,
  outerProps: OuterProps,
): InnerProps => {
  const store = useContext(context);
  const props = useMemo(() => (({} as any) as Store<State, InnerProps, OuterProps>), []);
  const createProps = useMemo(
    () =>
      selector.create<State, OuterProps, MapState, MapDispatch, InnerProps>(
        store.dispatch,
        mapStateToProps,
        mapDispatchToProps,
      ),
    [store],
  );

  props.outer = outerProps;
  props.state = store.getState();
  props.merge = useMemo(() => createProps(props.state, props.outer), [createProps, props.state, props.outer]);

  const [innerProps, updateProps] = useState(props.merge);

  props.inner = innerProps;

  useEffect(() => {
    // update the componentState only if props has changed
    updateProps(props.merge);
  }, [props.outer]);

  useEffect(() => {
    // update the componentState only if storeState has changed
    const subscription = store.subscribe((newState) => {
      if (!isStrictEqual(props.state, newState)) {
        const newProps = createProps(newState, props.outer);

        if (!isStrictEqual(props.inner, newProps)) {
          updateProps(newProps);
        }
      }
    });

    return () => subscription.unsubscribe();
  }, [store]);

  return props.inner;
};

export default useStore;
