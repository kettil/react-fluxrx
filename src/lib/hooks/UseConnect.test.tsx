import React, { createContext } from 'react';
import renderer from 'react-test-renderer';
import { StoreType } from '../types';
import useStore from './useConnect';

const oldState = { value: 42 };

const unsubscribe = jest.fn();
const subscribe = jest.fn().mockReturnValue({ unsubscribe });
const dispatch = jest.fn();
const getState = jest.fn();

const mapStateToProps = (state: any) => state;
const mapDispatchToProps = (dispatcher: any) => ({ onClick: () => dispatcher('add') });

describe('Check the useStore hook', () => {
  let results: any[];
  let context: React.Context<StoreType<any, any>>;
  let Component: React.FunctionComponent<{ type: string }>;
  let root: renderer.ReactTestRenderer | undefined;

  beforeEach(() => {
    getState.mockReturnValue(oldState);

    root = undefined;
    results = [];
    context = createContext<StoreType<any>>({ dispatch, getState, subscribe });

    Component = (props) => {
      const state = useStore(context, mapStateToProps, mapDispatchToProps, props);

      results.push(state);

      return React.createElement<any>('input', state);
    };
  });

  test('it should be return the element with merge props when the element is created', () => {
    renderer.act(() => {
      root = renderer.create(<Component type="text" />);
    });

    expect(root!.toJSON()).toMatchSnapshot('json');
    expect(results).toMatchSnapshot('results');
    expect(results.length).toBe(1);
    expect(dispatch).toHaveBeenCalledTimes(0);
    expect(getState).toHaveBeenCalledTimes(1);
    expect(subscribe).toHaveBeenCalledTimes(1);
    expect(unsubscribe).toHaveBeenCalledTimes(0);
    expect(results.length).toBe(1);

    renderer.act(() => {
      root!.unmount();
    });

    expect(unsubscribe).toHaveBeenCalledTimes(1);
  });

  test('it should be the element is recreated when element is updated with new props', () => {
    renderer.act(() => {
      root = renderer.create(<Component type="text" />);
    });

    expect(root!.toJSON()).toMatchSnapshot('json');

    renderer.act(() => {
      root!.update(<Component type="number" />);
    });

    expect(root!.toJSON()).toMatchSnapshot('json');
    expect(results).toMatchSnapshot('results');
    expect(results.length).toBe(3);
    expect(results[0] === results[1]).toBe(true);
    expect(results[1] === results[2]).toBe(false);
    expect(dispatch).toHaveBeenCalledTimes(0);
    expect(getState).toHaveBeenCalledTimes(3);
    expect(subscribe).toHaveBeenCalledTimes(1);
    expect(unsubscribe).toHaveBeenCalledTimes(0);
  });

  test('it should be the element is recreated when element is updated with same props', () => {
    renderer.act(() => {
      root = renderer.create(<Component type="text" />);
    });

    expect(root!.toJSON()).toMatchSnapshot('json');

    renderer.act(() => {
      root!.update(<Component type="text" />);
    });

    expect(root!.toJSON()).toMatchSnapshot('json');
    expect(results).toMatchSnapshot('results');
    expect(results.length).toBe(2);
    expect(results[0] === results[1]).toBe(true);
    expect(dispatch).toHaveBeenCalledTimes(0);
    expect(getState).toHaveBeenCalledTimes(2);
    expect(subscribe).toHaveBeenCalledTimes(1);
    expect(unsubscribe).toHaveBeenCalledTimes(0);
  });

  test('it should be the element iis recreated when the state is updated (new value)', () => {
    renderer.act(() => {
      root = renderer.create(<Component type="text" />);
    });

    expect(root!.toJSON()).toMatchSnapshot('json');
    expect(subscribe).toHaveBeenCalledTimes(1);
    expect(subscribe).toHaveBeenCalledWith(expect.any(Function));

    const newState = { value: 13 };
    getState.mockReturnValue(newState);

    renderer.act(() => {
      subscribe.mock.calls[0][0](newState);
    });

    expect(root!.toJSON()).toMatchSnapshot('json');
    expect(results).toMatchSnapshot('results');
    expect(results.length).toBe(2);
    expect(results[0] === results[1]).toBe(false);
    expect(dispatch).toHaveBeenCalledTimes(0);
    expect(getState).toHaveBeenCalledTimes(2);
    expect(subscribe).toHaveBeenCalledTimes(1);
    expect(unsubscribe).toHaveBeenCalledTimes(0);
  });

  test('it should be the element iis recreated when the state is updated (same value but diff reference)', () => {
    renderer.act(() => {
      root = renderer.create(<Component type="text" />);
    });

    expect(root!.toJSON()).toMatchSnapshot('json');
    expect(subscribe).toHaveBeenCalledTimes(1);
    expect(subscribe).toHaveBeenCalledWith(expect.any(Function));

    renderer.act(() => {
      subscribe.mock.calls[0][0]({ value: 42 });
    });

    expect(root!.toJSON()).toMatchSnapshot('json');
    expect(results).toMatchSnapshot('results');
    expect(results.length).toBe(1);
    expect(dispatch).toHaveBeenCalledTimes(0);
    expect(getState).toHaveBeenCalledTimes(1);
    expect(subscribe).toHaveBeenCalledTimes(1);
    expect(unsubscribe).toHaveBeenCalledTimes(0);
  });

  test('it should be the element iis recreated when the state is updated (same reference)', () => {
    renderer.act(() => {
      root = renderer.create(<Component type="text" />);
    });

    expect(root!.toJSON()).toMatchSnapshot('json');
    expect(subscribe).toHaveBeenCalledTimes(1);
    expect(subscribe).toHaveBeenCalledWith(expect.any(Function));

    renderer.act(() => {
      subscribe.mock.calls[0][0](oldState);
    });

    expect(root!.toJSON()).toMatchSnapshot('json');
    expect(results).toMatchSnapshot('results');
    expect(results.length).toBe(1);
    expect(dispatch).toHaveBeenCalledTimes(0);
    expect(getState).toHaveBeenCalledTimes(1);
    expect(subscribe).toHaveBeenCalledTimes(1);
    expect(unsubscribe).toHaveBeenCalledTimes(0);
  });
});
