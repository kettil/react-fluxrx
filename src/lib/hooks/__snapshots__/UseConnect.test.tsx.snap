// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check the useConnect hook it should be return the element with merge props when the element is created: json 1`] = `
<input
  onClick={[Function]}
  type="text"
  value={42}
/>
`;

exports[`Check the useConnect hook it should be return the element with merge props when the element is created: results 1`] = `
Array [
  Object {
    "onClick": [Function],
    "type": "text",
    "value": 42,
  },
]
`;

exports[`Check the useConnect hook it should be the element is recreated when element is updated with new props: json 1`] = `
<input
  onClick={[Function]}
  type="text"
  value={42}
/>
`;

exports[`Check the useConnect hook it should be the element is recreated when element is updated with new props: json 2`] = `
<input
  onClick={[Function]}
  type="number"
  value={42}
/>
`;

exports[`Check the useConnect hook it should be the element is recreated when element is updated with new props: results 1`] = `
Array [
  Object {
    "onClick": [Function],
    "type": "text",
    "value": 42,
  },
  Object {
    "onClick": [Function],
    "type": "text",
    "value": 42,
  },
  Object {
    "onClick": [Function],
    "type": "number",
    "value": 42,
  },
]
`;

exports[`Check the useConnect hook it should be the element is recreated when element is updated with same props: json 1`] = `
<input
  onClick={[Function]}
  type="text"
  value={42}
/>
`;

exports[`Check the useConnect hook it should be the element is recreated when element is updated with same props: json 2`] = `
<input
  onClick={[Function]}
  type="text"
  value={42}
/>
`;

exports[`Check the useConnect hook it should be the element is recreated when element is updated with same props: results 1`] = `
Array [
  Object {
    "onClick": [Function],
    "type": "text",
    "value": 42,
  },
  Object {
    "onClick": [Function],
    "type": "text",
    "value": 42,
  },
]
`;

exports[`Check the useConnect hook it should be the element is recreated when the state is updated (new value): json 1`] = `
<input
  onClick={[Function]}
  type="text"
  value={42}
/>
`;

exports[`Check the useConnect hook it should be the element is recreated when the state is updated (new value): json 2`] = `
<input
  onClick={[Function]}
  type="text"
  value={13}
/>
`;

exports[`Check the useConnect hook it should be the element is recreated when the state is updated (new value): results 1`] = `
Array [
  Object {
    "onClick": [Function],
    "type": "text",
    "value": 42,
  },
  Object {
    "onClick": [Function],
    "type": "text",
    "value": 13,
  },
]
`;

exports[`Check the useConnect hook it should be the element is recreated when the state is updated (same reference): json 1`] = `
<input
  onClick={[Function]}
  type="text"
  value={42}
/>
`;

exports[`Check the useConnect hook it should be the element is recreated when the state is updated (same reference): json 2`] = `
<input
  onClick={[Function]}
  type="text"
  value={42}
/>
`;

exports[`Check the useConnect hook it should be the element is recreated when the state is updated (same reference): results 1`] = `
Array [
  Object {
    "onClick": [Function],
    "type": "text",
    "value": 42,
  },
]
`;

exports[`Check the useConnect hook it should be the element is recreated when the state is updated (same value but diff reference): json 1`] = `
<input
  onClick={[Function]}
  type="text"
  value={42}
/>
`;

exports[`Check the useConnect hook it should be the element is recreated when the state is updated (same value but diff reference): json 2`] = `
<input
  onClick={[Function]}
  type="text"
  value={42}
/>
`;

exports[`Check the useConnect hook it should be the element is recreated when the state is updated (same value but diff reference): results 1`] = `
Array [
  Object {
    "onClick": [Function],
    "type": "text",
    "value": 42,
  },
]
`;
